File: src/frontend/src/app/services/Dtos/Envelope.ts
import { ErrorDto } from "./ErrorDto";

export class UntypedEnvelope {
    isSuccessful: boolean = false;
    error: ErrorDto | undefined;
    at: string;

    constructor() {
        this.isSuccessful = true;
        this.at = new Date().toUTCString();
        this.error = undefined;
    }

    static AsFailure(errorCode: number): UntypedEnvelope {
        const envelope = new UntypedEnvelope();
        envelope.isSuccessful = false;
        envelope.error = new ErrorDto();
        envelope.error.message = 'Seomething went wrong'; 
        envelope.error.code = errorCode;
        return envelope;
    }
}

export class Envelope<T> extends UntypedEnvelope {
    payload: T;

    constructor(payload: T) {
        super();
        this.payload = payload;
    }
}

File: src/backend/NewsTrack.WebApi/Dtos/Envelope.cs
using System;

namespace NewsTrack.WebApi.Dtos
{
    public class Envelope<T> : Envelope
        where T: class
    {
        public T Payload { get; set; }
    }

    public class Envelope
    {
        public bool IsSuccessful { get; set; }
        public DateTime At { get; set; }
        public Error Error { get; set; }
        
        public Envelope()
        {
            At = DateTime.UtcNow;
        }

        public static Envelope AsFailure(uint errorCode) => Envelope.AsFailure(null, errorCode);

        public static Envelope AsFailure(string errorMessage, uint errorCode = 0)
        {
            return new Envelope
            {
                IsSuccessful = false,
                Error = new Error
                {
                    Code = errorCode,
                    Message = errorMessage
                }
            };
        }

        public static Envelope AsSuccess() => new Envelope { IsSuccessful = true };
    }
}